File: 1499/solution.go
package _1499

import "math"

func findMaxValueOfEquation(points [][]int, k int) int {
	res := math.MinInt
	q := [][]int{}
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	for _, point := range points {
		x, y := point[0], point[1]
		for len(q) > 0 && x-q[0][1] > k {
			q = q[1:]
		}
		if len(q) > 0 {
			res = max(res, x+y+q[0][0])
		}
		for len(q) > 0 && y-x >= q[len(q)-1][0] {
			q = q[:len(q)-1]
		}
		q = append(q, []int{y - x, x})
	}
	return res
}

File: 1499/Solution.java
import java.util.Deque;
import java.util.LinkedList;

class Solution {
    public int findMaxValueOfEquation(int[][] points, int k) {
        int res = Integer.MIN_VALUE;
        Deque<int[]> queue = new LinkedList<>();
        for (int[] point: points) {
            int x = point[0], y = point[1];
            while (queue.size() > 0 && x - queue.getFirst()[1] > k) {
                queue.removeFirst();
            }
            if (queue.size() > 0) {
                res = Math.max(res, x + y + queue.getFirst()[0]);
            }
            while (queue.size() > 0 && y - x >= queue.getLast()[0]) {
                queue.removeLast();
            }
            queue.add(new int[]{y - x, x});
        }
        return res;
    }
}